Enable BloomFilter for _id of non-datastream indices

dnhatn opened the pull request:
This PR adds BloomFilter to Elasticsearch and enables it for the `_id` field of non-data stream indices. BloomFilter should speed up the performance of mget and update requests at a small expense of refresh, merge, and storage.

- Fork Lucene's BloomFilter to support an on-disk format and integrate with CircuitBreaker (i.e., ensure no OOM when building BloomFilter). We will also have a special treatment to skip reading BloomFilter for the archive tier. See:  https://github.com/elastic/elasticsearch/pull/88409/commits/cbb0800a359ca82e05c5d1a53848cc5676c76bbe.

- Enable BloomFilter for `_id` of non-data-stream indices: https://github.com/elastic/elasticsearch/pull/88409/commits/bc8c0d323185f7398f22672de40c937031cd4d41#diff-5779b7efbee3f9220189f8bdf03991e6f23cdac6506c5924e31231d8d0b0314aR52

If we are okay with this approach, I will split this PR into two smaller PRs.

### Benchmark results:

<details>
<summary> http_logs/update: 30% faster</summary>

```
|                                                        Metric |   Task |        Baseline |       Contender |       Diff |   Unit |   Diff % |
|--------------------------------------------------------------:|-------:|----------------:|----------------:|-----------:|-------:|---------:|
|                    Cumulative indexing time of primary shards |        |   204.488       |   134.085       |  -70.4033  |    min |  -34.43% |
|             Min cumulative indexing time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|          Median cumulative indexing time across primary shard |        |     5.12353     |     4.30781     |   -0.81573 |    min |  -15.92% |
|             Max cumulative indexing time across primary shard |        |   173.89        |   108.429       |  -65.4613  |    min |  -37.65% |
|           Cumulative indexing throttle time of primary shards |        |     0           |     0           |    0       |    min |    0.00% |
|    Min cumulative indexing throttle time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
| Median cumulative indexing throttle time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|    Max cumulative indexing throttle time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|                       Cumulative merge time of primary shards |        |    51.1106      |    53.4685      |    2.35792 |    min |   +4.61% |
|                      Cumulative merge count of primary shards |        |    39           |    38           |   -1       |        |   -2.56% |
|                Min cumulative merge time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|             Median cumulative merge time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|                Max cumulative merge time across primary shard |        |    51.0884      |    53.3766      |    2.28818 |    min |   +4.48% |
|              Cumulative merge throttle time of primary shards |        |    33.221       |    32.3083      |   -0.91268 |    min |   -2.75% |
|       Min cumulative merge throttle time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|    Median cumulative merge throttle time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|       Max cumulative merge throttle time across primary shard |        |    33.221       |    32.3083      |   -0.91268 |    min |   -2.75% |
|                     Cumulative refresh time of primary shards |        |     3.9797      |     4.25122     |    0.27152 |    min |   +6.82% |
|                    Cumulative refresh count of primary shards |        |   240           |   236           |   -4       |        |   -1.67% |
|              Min cumulative refresh time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|           Median cumulative refresh time across primary shard |        |     0.18635     |     0.197367    |    0.01102 |    min |   +5.91% |
|              Max cumulative refresh time across primary shard |        |     2.83907     |     3.11542     |    0.27635 |    min |   +9.73% |
|                       Cumulative flush time of primary shards |        |    19.2593      |    20.9628      |    1.70352 |    min |   +8.85% |
|                      Cumulative flush count of primary shards |        |   100           |   100           |    0       |        |    0.00% |
|                Min cumulative flush time across primary shard |        |     0           |     0           |    0       |    min |    0.00% |
|             Median cumulative flush time across primary shard |        |     0.553225    |     0.669417    |    0.11619 |    min |  +21.00% |
|                Max cumulative flush time across primary shard |        |    15.857       |    17.0961      |    1.23905 |    min |   +7.81% |
|                                       Total Young Gen GC time |        |   214.232       |   221.676       |    7.444   |      s |   +3.47% |
|                                      Total Young Gen GC count |        |  1510           |  1510           |    0       |        |    0.00% |
|                                         Total Old Gen GC time |        |     0           |     0           |    0       |      s |    0.00% |
|                                        Total Old Gen GC count |        |     0           |     0           |    0       |        |    0.00% |
|                                                    Store size |        |    15.5356      |    17.207       |    1.6714  |     GB |  +10.76% |
|                                                 Translog size |        |     4.09782e-07 |     4.09782e-07 |    0       |     GB |    0.00% |
|                                        Heap used for segments |        |     0           |     0           |    0       |     MB |    0.00% |
|                                      Heap used for doc values |        |     0           |     0           |    0       |     MB |    0.00% |
|                                           Heap used for terms |        |     0           |     0           |    0       |     MB |    0.00% |
|                                           Heap used for norms |        |     0           |     0           |    0       |     MB |    0.00% |
|                                          Heap used for points |        |     0           |     0           |    0       |     MB |    0.00% |
|                                   Heap used for stored fields |        |     0           |     0           |    0       |     MB |    0.00% |
|                                                 Segment count |        |   136           |   129           |   -7       |        |   -5.15% |
|                                   Total Ingest Pipeline count |        |     0           |     0           |    0       |        |    0.00% |
|                                    Total Ingest Pipeline time |        |     0           |     0           |    0       |     ms |    0.00% |
|                                  Total Ingest Pipeline failed |        |     0           |     0           |    0       |        |    0.00% |
|                                                Min Throughput | update | 17222.7         | 19471           | 2248.38    | docs/s |  +13.05% |
|                                               Mean Throughput | update | 25597.6         | 28594.9         | 2997.25    | docs/s |  +11.71% |
|                                             Median Throughput | update | 23691.3         | 26620.9         | 2929.63    | docs/s |  +12.37% |
|                                                Max Throughput | update | 38246.6         | 41607.9         | 3361.28    | docs/s |   +8.79% |
|                                       50th percentile latency | update |   102.637       |    80.426       |  -22.2111  |     ms |  -21.64% |
|                                       90th percentile latency | update |   186.097       |   158.1         |  -27.9972  |     ms |  -15.04% |
|                                       99th percentile latency | update |   356.333       |   329.398       |  -26.9353  |     ms |   -7.56% |
|                                     99.9th percentile latency | update |  1038.69        |  1061.98        |   23.2905  |     ms |   +2.24% |
|                                    99.99th percentile latency | update |  8934.49        |  9564.69        |  630.199   |     ms |   +7.05% |
|                                      100th percentile latency | update | 15070.3         | 17995.3         | 2925.01    |     ms |  +19.41% |
|                                  50th percentile service time | update |   102.637       |    80.426       |  -22.2111  |     ms |  -21.64% |
|                                  90th percentile service time | update |   186.097       |   158.1         |  -27.9972  |     ms |  -15.04% |
|                                  99th percentile service time | update |   356.333       |   329.398       |  -26.9353  |     ms |   -7.56% |
|                                99.9th percentile service time | update |  1038.69        |  1061.98        |   23.2905  |     ms |   +2.24% |
|                               99.99th percentile service time | update |  8934.49        |  9564.69        |  630.199   |     ms |   +7.05% |
|                                 100th percentile service time | update | 15070.3         | 17995.3         | 2925.01    |     ms |  +19.41% |
|                                                    error rate | update |     0           |     0           |    0       |      % |    0.00% |
```
</details>

<details> 
<summary>geonames/append-fast-with-conflicts: 11% faster</summary>

```
|                                                        Metric |         Task |        Baseline |       Contender |        Diff |   Unit |   Diff % |
|--------------------------------------------------------------:|-------------:|----------------:|----------------:|------------:|-------:|---------:|
|                    Cumulative indexing time of primary shards |              |    11.5674      |    10.2602      |    -1.30725 |    min |  -11.30% |
|             Min cumulative indexing time across primary shard |              |     1.79325     |     1.58932     |    -0.20393 |    min |  -11.37% |
|          Median cumulative indexing time across primary shard |              |     1.92863     |     1.69722     |    -0.23141 |    min |  -12.00% |
|             Max cumulative indexing time across primary shard |              |     2.11288     |     1.84133     |    -0.27155 |    min |  -12.85% |
|           Cumulative indexing throttle time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|    Min cumulative indexing throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
| Median cumulative indexing throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|    Max cumulative indexing throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                       Cumulative merge time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|                      Cumulative merge count of primary shards |              |     0           |     0           |     0       |        |    0.00% |
|                Min cumulative merge time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|             Median cumulative merge time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                Max cumulative merge time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|              Cumulative merge throttle time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|       Min cumulative merge throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|    Median cumulative merge throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|       Max cumulative merge throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                     Cumulative refresh time of primary shards |              |     2.97885     |     3.04907     |     0.07022 |    min |   +2.36% |
|                    Cumulative refresh count of primary shards |              |    51           |    53           |     2       |        |   +3.92% |
|              Min cumulative refresh time across primary shard |              |     0.459       |     0.45675     |    -0.00225 |    min |   -0.49% |
|           Median cumulative refresh time across primary shard |              |     0.49        |     0.508175    |     0.01818 |    min |   +3.71% |
|              Max cumulative refresh time across primary shard |              |     0.554917    |     0.54805     |    -0.00687 |    min |   -1.24% |
|                       Cumulative flush time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|                      Cumulative flush count of primary shards |              |     0           |     0           |     0       |        |    0.00% |
|                Min cumulative flush time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|             Median cumulative flush time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                Max cumulative flush time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                                       Total Young Gen GC time |              |     2.623       |     2.864       |     0.241   |      s |   +9.19% |
|                                      Total Young Gen GC count |              |    60           |    65           |     5       |        |   +8.33% |
|                                         Total Old Gen GC time |              |     0           |     0           |     0       |      s |    0.00% |
|                                        Total Old Gen GC count |              |     0           |     0           |     0       |        |    0.00% |
|                                                    Store size |              |     2.62687     |     2.65214     |     0.02527 |     GB |   +0.96% |
|                                                 Translog size |              |     3.07336e-07 |     3.07336e-07 |     0       |     GB |    0.00% |
|                                        Heap used for segments |              |     0           |     0           |     0       |     MB |    0.00% |
|                                      Heap used for doc values |              |     0           |     0           |     0       |     MB |    0.00% |
|                                           Heap used for terms |              |     0           |     0           |     0       |     MB |    0.00% |
|                                           Heap used for norms |              |     0           |     0           |     0       |     MB |    0.00% |
|                                          Heap used for points |              |     0           |     0           |     0       |     MB |    0.00% |
|                                   Heap used for stored fields |              |     0           |     0           |     0       |     MB |    0.00% |
|                                                 Segment count |              |    62           |    65           |     3       |        |   +4.84% |
|                                   Total Ingest Pipeline count |              |     0           |     0           |     0       |        |    0.00% |
|                                    Total Ingest Pipeline time |              |     0           |     0           |     0       |     ms |    0.00% |
|                                  Total Ingest Pipeline failed |              |     0           |     0           |     0       |        |    0.00% |
|                                                Min Throughput | index-update | 55712.1         | 58663.4         |  2951.27    | docs/s |   +5.30% |
|                                               Mean Throughput | index-update | 59827.6         | 63974.9         |  4147.34    | docs/s |   +6.93% |
|                                             Median Throughput | index-update | 60077.9         | 64345.7         |  4267.81    | docs/s |   +7.10% |
|                                                Max Throughput | index-update | 62332.9         | 65957.4         |  3624.41    | docs/s |   +5.81% |
|                                       50th percentile latency | index-update |    18.2543      |    17.7749      |    -0.47942 |     ms |   -2.63% |
|                                       90th percentile latency | index-update |    25.993       |    25.738       |    -0.25502 |     ms |   -0.98% |
|                                       99th percentile latency | index-update |    96.5377      |   105.799       |     9.26124 |     ms |   +9.59% |
|                                     99.9th percentile latency | index-update |  2374.35        |   655.37        | -1718.98    |     ms |  -72.40% |
|                                      100th percentile latency | index-update | 11559.5         | 14077.1         |  2517.68    |     ms |  +21.78% |
|                                  50th percentile service time | index-update |    18.2543      |    17.7749      |    -0.47942 |     ms |   -2.63% |
|                                  90th percentile service time | index-update |    25.993       |    25.738       |    -0.25502 |     ms |   -0.98% |
|                                  99th percentile service time | index-update |    96.5377      |   105.799       |     9.26124 |     ms |   +9.59% |
|                                99.9th percentile service time | index-update |  2374.35        |   655.37        | -1718.98    |     ms |  -72.40% |
|                                 100th percentile service time | index-update | 11559.5         | 14077.1         |  2517.68    |     ms |  +21.78% |
|                                                    error rate | index-update |     0           |     0           |     0       |      % |    0.00% |
```
</details> 


<details> 
<summary>http_logs/append-no-conflicts-index-only: 4% faster</summary>

```
|                                                        Metric |         Task |         Baseline |        Contender |       Diff |   Unit |   Diff % |
|--------------------------------------------------------------:|-------------:|-----------------:|-----------------:|-----------:|-------:|---------:|
|                    Cumulative indexing time of primary shards |              |     94.639       |     90.2686      |   -4.3704  |    min |   -4.62% |
|             Min cumulative indexing time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|          Median cumulative indexing time across primary shard |              |      0.854917    |      0.799267    |   -0.05565 |    min |   -6.51% |
|             Max cumulative indexing time across primary shard |              |     14.4983      |     13.7474      |   -0.75092 |    min |   -5.18% |
|           Cumulative indexing throttle time of primary shards |              |      0           |      0           |    0       |    min |    0.00% |
|    Min cumulative indexing throttle time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
| Median cumulative indexing throttle time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|    Max cumulative indexing throttle time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|                       Cumulative merge time of primary shards |              |     10.655       |     11.2547      |    0.59968 |    min |   +5.63% |
|                      Cumulative merge count of primary shards |              |     29           |     28           |   -1       |        |   -3.45% |
|                Min cumulative merge time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|             Median cumulative merge time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|                Max cumulative merge time across primary shard |              |      2.39942     |      2.57625     |    0.17683 |    min |   +7.37% |
|              Cumulative merge throttle time of primary shards |              |      4.43008     |      4.80807     |    0.37798 |    min |   +8.53% |
|       Min cumulative merge throttle time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|    Median cumulative merge throttle time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|       Max cumulative merge throttle time across primary shard |              |      0.992817    |      1.1482      |    0.15538 |    min |  +15.65% |
|                     Cumulative refresh time of primary shards |              |      8.6277      |      9.14297     |    0.51527 |    min |   +5.97% |
|                    Cumulative refresh count of primary shards |              |    467           |    456           |  -11       |        |   -2.36% |
|              Min cumulative refresh time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|           Median cumulative refresh time across primary shard |              |      0.135583    |      0.146125    |    0.01054 |    min |   +7.78% |
|              Max cumulative refresh time across primary shard |              |      1.55802     |      1.69002     |    0.132   |    min |   +8.47% |
|                       Cumulative flush time of primary shards |              |      8.83267     |      9.4904      |    0.65773 |    min |   +7.45% |
|                      Cumulative flush count of primary shards |              |    123           |    123           |    0       |        |    0.00% |
|                Min cumulative flush time across primary shard |              |      0           |      0           |    0       |    min |    0.00% |
|             Median cumulative flush time across primary shard |              |      0.0178583   |      0.017375    |   -0.00048 |    min |   -2.71% |
|                Max cumulative flush time across primary shard |              |      2.29292     |      2.23348     |   -0.05943 |    min |   -2.59% |
|                                       Total Young Gen GC time |              |      8.215       |      8.511       |    0.296   |      s |   +3.60% |
|                                      Total Young Gen GC count |              |    757           |    744           |  -13       |        |   -1.72% |
|                                         Total Old Gen GC time |              |      0           |      0           |    0       |      s |    0.00% |
|                                        Total Old Gen GC count |              |      0           |      0           |    0       |        |    0.00% |
|                                                    Store size |              |     16.469       |     16.8836      |    0.41457 |     GB |   +2.52% |
|                                                 Translog size |              |      2.04891e-06 |      2.04891e-06 |    0       |     GB |    0.00% |
|                                        Heap used for segments |              |      0           |      0           |    0       |     MB |    0.00% |
|                                      Heap used for doc values |              |      0           |      0           |    0       |     MB |    0.00% |
|                                           Heap used for terms |              |      0           |      0           |    0       |     MB |    0.00% |
|                                           Heap used for norms |              |      0           |      0           |    0       |     MB |    0.00% |
|                                          Heap used for points |              |      0           |      0           |    0       |     MB |    0.00% |
|                                   Heap used for stored fields |              |      0           |      0           |    0       |     MB |    0.00% |
|                                                 Segment count |              |    278           |    296           |   18       |        |   +6.47% |
|                                   Total Ingest Pipeline count |              |      0           |      0           |    0       |        |    0.00% |
|                                    Total Ingest Pipeline time |              |      0           |      0           |    0       |     ms |    0.00% |
|                                  Total Ingest Pipeline failed |              |      0           |      0           |    0       |        |    0.00% |
|                                                Min Throughput | index-append | 128322           | 128995           |  673.199   | docs/s |   +0.52% |
|                                               Mean Throughput | index-append | 131048           | 134264           | 3216.09    | docs/s |   +2.45% |
|                                             Median Throughput | index-append | 131401           | 134944           | 3542.48    | docs/s |   +2.70% |
|                                                Max Throughput | index-append | 132607           | 138927           | 6319.52    | docs/s |   +4.77% |
|                                       50th percentile latency | index-append |     11.7579      |     11.7469      |   -0.011   |     ms |   -0.09% |
|                                       90th percentile latency | index-append |     14.3791      |     14.6843      |    0.30518 |     ms |   +2.12% |
|                                       99th percentile latency | index-append |     23.3543      |     23.7754      |    0.4211  |     ms |   +1.80% |
|                                     99.9th percentile latency | index-append |    230.45        |    248.695       |   18.245   |     ms |   +7.92% |
|                                    99.99th percentile latency | index-append |   6502.86        |   7400.32        |  897.454   |     ms |  +13.80% |
|                                      100th percentile latency | index-append |  10307           |  10792.7         |  485.707   |     ms |   +4.71% |
|                                  50th percentile service time | index-append |     11.7579      |     11.7469      |   -0.011   |     ms |   -0.09% |
|                                  90th percentile service time | index-append |     14.3791      |     14.6843      |    0.30518 |     ms |   +2.12% |
|                                  99th percentile service time | index-append |     23.3543      |     23.7754      |    0.4211  |     ms |   +1.80% |
|                                99.9th percentile service time | index-append |    230.45        |    248.695       |   18.245   |     ms |   +7.92% |
|                               99.99th percentile service time | index-append |   6502.86        |   7400.32        |  897.454   |     ms |  +13.80% |
|                                 100th percentile service time | index-append |  10307           |  10792.7         |  485.707   |     ms |   +4.71% |
|                                                    error rate | index-append |      0.000463411 |      0.00327129  |    0.00281 |      % | +605.91% |

```
</details>


<details>
<summary>geonames/append-no-conflicts-index-only: 4% slower</summary>

```
|                                                        Metric |         Task |        Baseline |       Contender |        Diff |   Unit |   Diff % |
|--------------------------------------------------------------:|-------------:|----------------:|----------------:|------------:|-------:|---------:|
|                    Cumulative indexing time of primary shards |              |     9.71393     |    10.1481      |     0.43418 |    min |   +4.47% |
|             Min cumulative indexing time across primary shard |              |     1.79678     |     1.87212     |     0.07533 |    min |   +4.19% |
|          Median cumulative indexing time across primary shard |              |     2.02563     |     2.11442     |     0.08878 |    min |   +4.38% |
|             Max cumulative indexing time across primary shard |              |     2.05453     |     2.13473     |     0.0802  |    min |   +3.90% |
|           Cumulative indexing throttle time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|    Min cumulative indexing throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
| Median cumulative indexing throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|    Max cumulative indexing throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                       Cumulative merge time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|                      Cumulative merge count of primary shards |              |     0           |     0           |     0       |        |    0.00% |
|                Min cumulative merge time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|             Median cumulative merge time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                Max cumulative merge time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|              Cumulative merge throttle time of primary shards |              |     0           |     0           |     0       |    min |    0.00% |
|       Min cumulative merge throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|    Median cumulative merge throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|       Max cumulative merge throttle time across primary shard |              |     0           |     0           |     0       |    min |    0.00% |
|                     Cumulative refresh time of primary shards |              |     1.9438      |     2.14998     |     0.20618 |    min |  +10.61% |
|                    Cumulative refresh count of primary shards |              |    54           |    52           |    -2       |        |   -3.70% |
|              Min cumulative refresh time across primary shard |              |     0.222633    |     0.266867    |     0.04423 |    min |  +19.87% |
|           Median cumulative refresh time across primary shard |              |     0.429133    |     0.476483    |     0.04735 |    min |  +11.03% |
|              Max cumulative refresh time across primary shard |              |     0.518733    |     0.602817    |     0.08408 |    min |  +16.21% |
|                       Cumulative flush time of primary shards |              |     0.79775     |     1.05328     |     0.25553 |    min |  +32.03% |
|                      Cumulative flush count of primary shards |              |     5           |     5           |     0       |        |    0.00% |
|                Min cumulative flush time across primary shard |              |     0.10165     |     0.144867    |     0.04322 |    min |  +42.52% |
|             Median cumulative flush time across primary shard |              |     0.142433    |     0.2163      |     0.07387 |    min |  +51.86% |
|                Max cumulative flush time across primary shard |              |     0.2252      |     0.297917    |     0.07272 |    min |  +32.29% |
|                                       Total Young Gen GC time |              |     1.032       |     1.135       |     0.103   |      s |   +9.98% |
|                                      Total Young Gen GC count |              |    48           |    51           |     3       |        |   +6.25% |
|                                         Total Old Gen GC time |              |     0           |     0           |     0       |      s |    0.00% |
|                                        Total Old Gen GC count |              |     0           |     0           |     0       |        |    0.00% |
|                                                    Store size |              |     2.63287     |     2.64842     |     0.01555 |     GB |   +0.59% |
|                                                 Translog size |              |     2.56114e-07 |     2.56114e-07 |     0       |     GB |    0.00% |
|                                        Heap used for segments |              |     0           |     0           |     0       |     MB |    0.00% |
|                                      Heap used for doc values |              |     0           |     0           |     0       |     MB |    0.00% |
|                                           Heap used for terms |              |     0           |     0           |     0       |     MB |    0.00% |
|                                           Heap used for norms |              |     0           |     0           |     0       |     MB |    0.00% |
|                                          Heap used for points |              |     0           |     0           |     0       |     MB |    0.00% |
|                                   Heap used for stored fields |              |     0           |     0           |     0       |     MB |    0.00% |
|                                                 Segment count |              |    84           |    83           |    -1       |        |   -1.19% |
|                                   Total Ingest Pipeline count |              |     0           |     0           |     0       |        |    0.00% |
|                                    Total Ingest Pipeline time |              |     0           |     0           |     0       |     ms |    0.00% |
|                                  Total Ingest Pipeline failed |              |     0           |     0           |     0       |        |    0.00% |
|                                                Min Throughput | index-append | 65646.2         | 63341.9         | -2304.33    | docs/s |   -3.51% |
|                                               Mean Throughput | index-append | 70545           | 66032.9         | -4512.07    | docs/s |   -6.40% |
|                                             Median Throughput | index-append | 70773.1         | 65718.9         | -5054.22    | docs/s |   -7.14% |
|                                                Max Throughput | index-append | 72382.5         | 68391.2         | -3991.26    | docs/s |   -5.51% |
|                                       50th percentile latency | index-append |    18.2187      |    18.9626      |     0.74386 |     ms |   +4.08% |
|                                       90th percentile latency | index-append |    22.3928      |    24.868       |     2.47521 |     ms |  +11.05% |
|                                       99th percentile latency | index-append |    72.5306      |    68.4397      |    -4.09095 |     ms |   -5.64% |
|                                     99.9th percentile latency | index-append |   260.41        |   463.791       |   203.381   |     ms |  +78.10% |
|                                      100th percentile latency | index-append |  9553.07        |  9428.72        |  -124.35    |     ms |   -1.30% |
|                                  50th percentile service time | index-append |    18.2187      |    18.9626      |     0.74386 |     ms |   +4.08% |
|                                  90th percentile service time | index-append |    22.3928      |    24.868       |     2.47521 |     ms |  +11.05% |
|                                  99th percentile service time | index-append |    72.5306      |    68.4397      |    -4.09095 |     ms |   -5.64% |
|                                99.9th percentile service time | index-append |   260.41        |   463.791       |   203.381   |     ms |  +78.10% |
|                                 100th percentile service time | index-append |  9553.07        |  9428.72        |  -124.35    |     ms |   -1.30% |
|                                                    error rate | index-append |     0           |     0           |     0       |      % |    0.00% |
```
</details>

dnhatn committed:
Add BloomFilterPostingsFormat

dnhatn committed:
Enable BloomFilter for _id of non-datastream indices

dnhatn commented:
run elasticsearch-ci/part-2

dnhatn commented:
run elasticsearch-ci/part-2

elasticsearchmachine commented:
Hi @dnhatn, I've created a changelog YAML for you.

dnhatn committed:
Update docs/changelog/88409.yaml

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Modifications copyright (C) 2021 Elasticsearch B.V.
```
```suggestion
 * Modifications copyright (C) 2022 Elasticsearch B.V.
```

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+import java.util.Objects;
+
+/**
+ * This implementation is forked from Lucene's BloomFilterPosting to support on-disk bloom filters.
```
Let's also contribute this change to the Lucene BloomFilteringPostingsFormat? (Though it might still make sense to maintain a fork to retain control on backward compatibility, since the Lucene format doesn't give any guarantee.)

dnhatn replied:
+1

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+ * offers "fast-fail" for reads in segments known to have no record of the key.
+ */
+public class BloomFilterPostingsFormat extends PostingsFormat {
+    static final String BLOOM_CODEC_NAME = "ESBloomFilter";
```
Let's version it so that we have namespace for new implementations in the future, e.g.
```suggestion
    static final String BLOOM_CODEC_NAME = "ES84BloomFilter";
```

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+ */
+public class BloomFilterPostingsFormat extends PostingsFormat {
+    static final String BLOOM_CODEC_NAME = "ESBloomFilter";
+    static final int VERSION_START = 1;
```
We usually start at 0, let's make it 0 for consistency?
```suggestion
    static final int VERSION_START = 0;
```

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+    @Override
+    public FieldsConsumer fieldsConsumer(SegmentWriteState state) throws IOException {
+        if (postingsFormat == null || bigArrays == null) {
+            assert false : "ESBloomFilterPostingsFormat was initialized with a wrong constructor";
```
nit: the message says `ESBloomFilterPostingsFormat` while the class is called `BloomFilterPostingsFormat`

dnhatn replied:
Should we rename this class to `ESBloomFilterPostingsFormat` instead?

dnhatn replied:
I renamed BloomFilterPostingsFormat to [ES84BloomFilterPostingsFormat](https://github.com/elastic/elasticsearch/pull/88409/commits/0fabf37e1973f9dedb4aff01b746c5341ea0c06f)

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+
+    @Override
+    public String toString() {
+        return "ESBloomFilterPostingsFormat(" + postingsFormat.getName() + ")";
```
likewise here

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+        return "ESBloomFilterPostingsFormat(" + postingsFormat.getName() + ")";
+    }
+
+    private static String metaField(SegmentInfo si, String segmentSuffix) {
```
did you mean `file` rather than `field` in the method name?

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+                        if (term == null) {
+                            break;
+                        }
+                        // Do we need to skip terms without postings?
```
It shouldn't be necessary, flush and merge guarantee that all terms would have postings when we get here

dnhatn replied:
ok, I removed this TODO

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+                    VERSION_CURRENT,
+                    state.segmentInfo.getId(),
+                    state.segmentSuffix
+                );
```
Can you also do a `CodecUtil.retrieveChecksum` on `indexIn` to verify that the checksum has a correct structure?

dnhatn replied:
added in https://github.com/elastic/elasticsearch/pull/88409/commits/62aedf97a1f8dfa5d6023eb3fd36ddf76af29c63

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+     * choice of hashing functions as the core function had to be adapted to work with BytesRefs with
+     * offsets and lengths rather than raw byte arrays.
+     */
+    private static int hashTerm(byte[] data, int seed, int offset, int len) {
```
Maybe switch to `StringHelper.murmurhash3_x86_32` instead, it should be both faster and simpler? (Though I agree with forking it so that we wouldn't break bw compat if this method changed in Lucene)

dnhatn committed:
Update server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

dnhatn committed:
Update server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

dnhatn committed:
Update server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

dnhatn committed:
Fix initialize toCloses size

dnhatn committed:
Remove TODO

dnhatn committed:
Use murmurhash3_x86_32

dnhatn committed:
typo: field -> file

dnhatn committed:
BloomFilterPostingsFormat -> ES84BloomFilterPostingsFormat

dnhatn committed:
retrieveChecksum

dnhatn commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,472 @@
+
+        FieldsWriter(PostingsFormat postingsFormat, SegmentWriteState state) throws IOException {
+            this.state = state;
+            final List<Closeable> toCloses = new ArrayList<>(1);
```
a left over, I fixed it in https://github.com/elastic/elasticsearch/pull/88409/commits/7ca0708fc2c27db468765073ed0e352bf4d2cb1e

dnhatn committed:
fix license

dnhatn committed:
Add index settings

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into bloom_filter

dnhatn committed:
fix test

dnhatn commented:
@jpountz Thank you for the review. I have addressed your comments and added an index setting in https://github.com/elastic/elasticsearch/pull/88409/commits/22c6a0e6bfbfd80d3f1079ba64b0ea8c42ad2960.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

dnhatn committed:
Merge branch 'main' into bloom_filter

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES84BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,479 @@
+
+        @Override
+        public void checkIntegrity() throws IOException {
+            termsReader.checkIntegrity();
```
Let's also do a `checksumEntireFile` on the file that holds the bloom filter? (No need to check the meta file since it's checked upon opening)

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES84BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,479 @@
+
+    static int hashTerm(BytesRef br) {
+        final int hash = murmurhash3_x86_32(br.bytes, br.offset, br.length, 0x9747b28c);
+        return hash < 0 ? -hash : hash;
```
This is fragile, as `- Integer.MIN_VALUE` is still a negative value (the min integer value is -2^31 but the max integer value is 2^31-1). Let's clear the sign bit instead? Ie. `return hash &0x7FFFFFFF`.

dnhatn replied:
Great catch 👍

jpountz commented on server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES84BloomFilterPostingsFormat.java:
```diff
@@ -0,0 +1,479 @@
+    }
+
+    /**
+     * Forked from Lucene's StringHelper#murmurhash3_x86_32
```
```suggestion
     * Forked from Lucene's StringHelper#murmurhash3_x86_32 so that changes to the Lucene implementation do not break the compatibility of this format.
```

jpountz commented on server/src/test/java/org/elasticsearch/index/codec/bloomfilter/ES84BloomFilterPostingsFormatTests.java:
```diff
@@ -0,0 +1,43 @@
+        assertThat(ES84BloomFilterPostingsFormat.numBytesForBloomFilter(16384), equalTo(2048));
+        assertThat(ES84BloomFilterPostingsFormat.numBytesForBloomFilter(16383), equalTo(2048));
+        assertThat(ES84BloomFilterPostingsFormat.numBytesForBloomFilter(Integer.MAX_VALUE), equalTo(1 << 28));
+    }
```
maybe also add a test that calls `hashTerm` on a couple fixed values and asserts on the result, so that we would detect unintended breaks of the hashing function, e.g. through mechanical refactorings?

dnhatn replied:
++, added tests in https://github.com/elastic/elasticsearch/pull/88409/commits/eb868eb2a50517f3d3a6e44064a9659f4d38b42d

jpountz reviewed (approved):
I left some suggestions but the change looks good to me.

dnhatn committed:
Update server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES84BloomFilterPostingsFormat.java

Co-authored-by: Adrien Grand <jpountz@gmail.com>

dnhatn committed:
rename to ES85

dnhatn committed:
Mark setting as deprecated_warning

dnhatn committed:
Add hash test

dnhatn committed:
fully check index file

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into bloom_filter

dnhatn commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -455,6 +455,17 @@ public final class IndexSettings {
+        true,
+        Setting.Property.Dynamic,
+        Setting.Property.IndexScope
+    );
```
I marked this settings `deprecated_warning` in https://github.com/elastic/elasticsearch/pull/88409/commits/a1bc3cfc2cd53ad450e20adc4d8ae778b0f07a48.

dnhatn committed:
Support multiple fields

dnhatn commented:
@jpountz I pushed https://github.com/elastic/elasticsearch/pull/88409/commits/85183414d944291ab7427159e2c57ee6d304a677 to share a single instance BloomFilterPostingsFormat between multiple fields. Can you take another look? Thank you!

dnhatn commented:
@jpountz Thanks for reviews.
